// wxpage/dist/wxpage.js
!(function(e, t) {
  for (var n in t) e[n] = t[n];
})(
  exports,
  (function(e) {
    var t = {};
    function n(r) {
      if (t[r]) return t[r].exports;
      var o = (t[r] = { i: r, l: !1, exports: {} });
      return e[r].call(o.exports, o, o.exports, n), (o.l = !0), o.exports;
    }
    return (
      (n.m = e),
      (n.c = t),
      (n.d = function(e, t, r) {
        n.o(e, t) || Object.defineProperty(e, t, { enumerable: !0, get: r });
      }),
      (n.r = function(e) {
        "undefined" != typeof Symbol &&
          Symbol.toStringTag &&
          Object.defineProperty(e, Symbol.toStringTag, { value: "Module" }),
          Object.defineProperty(e, "__esModule", { value: !0 });
      }),
      (n.t = function(e, t) {
        if ((1 & t && (e = n(e)), 8 & t)) return e;
        if (4 & t && "object" == typeof e && e && e.__esModule) return e;
        var r = Object.create(null);
        if (
          (n.r(r),
          Object.defineProperty(r, "default", { enumerable: !0, value: e }),
          2 & t && "string" != typeof e)
        )
          for (var o in e)
            n.d(
              r,
              o,
              function(t) {
                return e[t];
              }.bind(null, o)
            );
        return r;
      }),
      (n.n = function(e) {
        var t =
          e && e.__esModule
            ? function() {
                return e.default;
              }
            : function() {
                return e;
              };
        return n.d(t, "a", t), t;
      }),
      (n.o = function(e, t) {
        return Object.prototype.hasOwnProperty.call(e, t);
      }),
      (n.p = ""),
      n((n.s = 6))
    );
  })([
    function(e, t, n) {
      "use strict";
      Object.defineProperty(t, "__esModule", { value: !0 });
      var r = (function() {
        function e() {
          this.eventObject = {};
        }
        return (
          (e.prototype.on = function(e, t, n) {
            var r = this;
            return (
              void 0 === n && (n = !1),
              this.eventObject[e] || (this.eventObject[e] = []),
              this.eventObject[e].push({ handler: t, once: n }),
              function() {
                return r.off(e, t);
              }
            );
          }),
          (e.prototype.off = function(e, t) {
            var n = this;
            void 0 === e && (e = "all"),
              (Array.isArray(e)
                ? e
                : "all" === e
                ? Object.keys(this.eventObject)
                : [e]
              ).forEach(function(e) {
                if (t) {
                  var r = n.eventObject[e] || [],
                    o = [];
                  r.forEach(function(e) {
                    e.handler !== t && o.push(e);
                  }),
                    (n.eventObject[e] = o);
                } else n.eventObject[e] = [];
              });
          }),
          (e.prototype.emit = function(e) {
            for (var t = [], n = 1; n < arguments.length; n++)
              t[n - 1] = arguments[n];
            var r = this.eventObject[e] || [];
            r.forEach(function(n) {
              if (!n.once || !n.called) {
                n.called = !0;
                try {
                  n.handler && n.handler.apply(void 0, t);
                } catch (t) {
                  console.error(
                    e + "事件监听触发失败：",
                    t.stack || t.message || t
                  );
                }
              }
            });
          }),
          (e.prototype.assign = function(e) {
            var t = this;
            ["on", "off", "emit", "assign"].forEach(function(n) {
              e[n] = function() {
                for (var e = [], r = 0; r < arguments.length; r++)
                  e[r] = arguments[r];
                return t[n].apply(t, e);
              };
            });
          }),
          e
        );
      })();
      (t.message = new r()), (t.default = r);
    },
    function(e, t, n) {
      "use strict";
      var r =
        (this && this.__importDefault) ||
        function(e) {
          return e && e.__esModule ? e : { default: e };
        };
      Object.defineProperty(t, "__esModule", { value: !0 });
      var o = r(n(2)),
        a = new Date().getTime();
      console.info("当前sessionId为" + a);
      var u = function(e, t) {
        return t
          ? t.expr
            ? t.expr === a || new Date().getTime() < t.expr
              ? t.data
              : (wx.removeStorageSync("_cache_" + e), null)
            : t.data
          : null;
      };
      t.default = {
        set: function(e, t, n, r) {
          "function" === o.default.type(n)
            ? ((r = n), (n = "once"))
            : void 0 === n && (n = "once");
          var u = { expr: 0, data: t };
          if ("keep" === n) {
            var i = wx.getStorageSync("_cache_" + e);
            if (!i) return;
            n = i.expr || 0;
          } else "once" === n ? (n = a) : (n += new Date().getTime());
          (u.expr = n),
            r
              ? wx.setStorage({
                  data: u,
                  key: "_cache_" + e,
                  success: function() {
                    r();
                  },
                  fail: function(t) {
                    r(t || 'set "' + e + '" fail');
                  }
                })
              : wx.setStorageSync("_cache_" + e, u);
        },
        get: function(e, t) {
          if (!t) return u(e, wx.getStorageSync("_cache_" + e));
          wx.getStorage({
            key: "_cache_" + e,
            success: function(n) {
              t(u(e, n.data));
            },
            fail: function(n) {
              t(n || 'get "' + e + '" fail');
            }
          });
        },
        remove: function(e, t) {
          t
            ? wx.removeStorage({
                key: "_cache_" + e,
                success: function() {
                  t();
                },
                fail: function(n) {
                  t(n || 'remove "' + e + '" fail');
                }
              })
            : (wx.removeStorageSync("_cache_" + e), u(e, null));
        },
        resolve: u
      };
    },
    function(e, t, n) {
      "use strict";
      Object.defineProperty(t, "__esModule", { value: !0 });
      var r = function(e, t) {
          return e && e.hasOwnProperty(t);
        },
        o = function(e, t, n) {
          return (
            void 0 === t && (t = "&"),
            void 0 === n && (n = !1),
            e
              ? Object.keys(e)
                  .map(function(t) {
                    var r = e[t];
                    return t + "=" + (n ? r : encodeURIComponent(r));
                  })
                  .join(t)
              : ""
          );
        };
      t.default = {
        type: function(e) {
          if ("object" == typeof e) {
            if (null === e) return "null";
            if (void 0 === e) return "undefined";
            var t = /\[object (\w+)\]/u.exec(Object.prototype.toString.call(e));
            return t ? t[1].toLowerCase() : "";
          }
          return typeof e;
        },
        hasOwn: r,
        objEach: function(e, t) {
          if (e) for (var n in e) if (r(e, n) && t(n, !1 === e[n])) break;
        },
        mergeFun: function(e, t) {
          return function() {
            for (var n = [], r = 0; r < arguments.length; r++)
              n[r] = arguments[r];
            try {
              t && t.apply(this, n);
            } finally {
              e && e.apply(this, n);
            }
          };
        },
        lock: function(e) {
          var t;
          return function() {
            for (var n = [], r = 0; r < arguments.length; r++)
              n[r] = arguments[r];
            if (!t) {
              t = !0;
              var o = [].slice.call(n, 0);
              o.unshift(function() {
                t = !1;
              }),
                e.apply(this, o);
            }
          };
        },
        once: function(e, t) {
          var n;
          return function() {
            for (var r = [], o = 0; o < arguments.length; o++)
              r[o] = arguments[o];
            !n && e && ((n = !0), e.apply(t || null, r));
          };
        },
        queryParse: function(e, t) {
          if ((void 0 === t && (t = "&"), !e)) return {};
          var n = {},
            r = e ? e.split(t) : void 0;
          return (
            r &&
              r.length > 0 &&
              r.forEach(function(e) {
                var t = e.split("="),
                  r = t[0],
                  o = t[1];
                n[r] = o;
              }),
            n
          );
        },
        queryStringify: o,
        queryJoin: function(e, t, n) {
          void 0 === n && (n = !1);
          var r,
            a = o(t, "&", n);
          return a
            ? ((r = /[\?&]$/u.test(e) ? "" : -1 === e.indexOf("?") ? "?" : "&"),
              "" + e + r + a)
            : e;
        }
      };
    },
    function(e, t, n) {
      "use strict";
      Object.defineProperty(t, "__esModule", { value: !0 });
      var r = { nameResolve: function() {} };
      t.default = {
        set: function(e, t) {
          switch (e) {
            case "resolvePath":
              "function" == typeof t && (r.customRouteResolve = t);
              break;
            case "route":
              if ("string" == typeof t || Array.isArray(t)) {
                var n = "string" == typeof t ? [t] : t,
                  o = n[0];
                (n = n.map(function(e) {
                  return new RegExp(
                    "^" +
                      e
                        .replace(/^\/?/u, "/?")
                        .replace(/[\.]/gu, "\\.")
                        .replace("$page", "([\\w\\-]+)")
                  );
                })),
                  (r.routeResolve = function(e) {
                    return o.replace("$page", e);
                  }),
                  (r.nameResolve = function(e) {
                    var t = "";
                    return (
                      n.some(function(n) {
                        var r = n.exec(e);
                        return !!r && ((t = r[1]), !0);
                      }),
                      t
                    );
                  });
              } else console.error("不合理的route选项:", t);
              break;
            default:
              r[e] = t;
          }
        },
        get: function(e) {
          return r[e];
        }
      };
    },
    function(e, t, n) {
      "use strict";
      var r,
        o =
          (this && this.__extends) ||
          ((r = function(e, t) {
            return (r =
              Object.setPrototypeOf ||
              ({ __proto__: [] } instanceof Array &&
                function(e, t) {
                  e.__proto__ = t;
                }) ||
              function(e, t) {
                for (var n in t) t.hasOwnProperty(n) && (e[n] = t[n]);
              })(e, t);
          }),
          function(e, t) {
            function n() {
              this.constructor = e;
            }
            r(e, t),
              (e.prototype =
                null === t
                  ? Object.create(t)
                  : ((n.prototype = t.prototype), new n()));
          }),
        a =
          (this && this.__importDefault) ||
          function(e) {
            return e && e.__esModule ? e : { default: e };
          };
      Object.defineProperty(t, "__esModule", { value: !0 });
      var u,
        i,
        c,
        f = new ((function(e) {
          function t() {
            return (null !== e && e.apply(this, arguments)) || this;
          }
          return (
            o(t, e),
            (t.prototype.route = function(e, t, n) {
              c ||
                ((c = !0),
                clearTimeout(u),
                clearTimeout(i),
                (u = setTimeout(function() {
                  c = !1;
                }, 2e3)),
                this.emit("navigateTo", t.url),
                wx[e] && wx[e].apply(wx, n));
            }),
            (t.prototype.navigateTo = function(e) {
              for (var t = [], n = 1; n < arguments.length; n++)
                t[n - 1] = arguments[n];
              this.route("navigateTo", e, t);
            }),
            (t.prototype.reLaunch = function(e) {
              for (var t = [], n = 1; n < arguments.length; n++)
                t[n - 1] = arguments[n];
              this.route("reLaunch", e, t);
            }),
            (t.prototype.navigateBack = function(e) {
              wx.navigateBack.apply(wx, e);
            }),
            (t.prototype.switchTab = function(e) {
              for (var t = [], n = 1; n < arguments.length; n++)
                t[n - 1] = arguments[n];
              this.route("switchTab", e, t);
            }),
            (t.prototype.redirectTo = function(e) {
              for (var t = [], n = 1; n < arguments.length; n++)
                t[n - 1] = arguments[n];
              this.route("redirectTo", e, t);
            }),
            t
          );
        })(a(n(0)).default))();
      f.on("page:ready", function() {
        i = setTimeout(function() {
          c = !1;
        }, 100);
      }),
        (t.default = f);
    },
    function(e, t, n) {
      "use strict";
      var r =
        (this && this.__importDefault) ||
        function(e) {
          return e && e.__esModule ? e : { default: e };
        };
      Object.defineProperty(t, "__esModule", { value: !0 });
      var o,
        a = r(n(1)),
        u = r(n(3)),
        i = r(n(4)),
        c = {};
      t.queryParse = function(e, t) {
        if ((void 0 === t && (t = "&"), !e)) return {};
        var n = {},
          r = e ? e.split(t) : void 0;
        return (
          r &&
            r.length > 0 &&
            r.forEach(function(e) {
              var t = e.split("="),
                r = t[0],
                o = t[1];
              n[r] = o;
            }),
          n
        );
      };
      var f,
        l = function(e) {
          return function(t, n) {
            var r,
              o = t.split("?"),
              a = o[0],
              c = o[1];
            if (
              (/^[\w\-]+$/u.test(a) &&
                (r = (u.default.get("customRouteResolve") ||
                  u.default.get("routeResolve"))(a)),
              !r)
            )
              throw new Error("Invalid path:" + r);
            ((n = n || {}).url = r + (c ? "?" + c : "")),
              i.default[e](n, { url: n.url });
          };
        },
        s = l("navigateTo"),
        d = l("redirectTo"),
        p = l("switchTab"),
        h = l("reLaunch"),
        v = { navigate: s, redirect: d, switchTab: p, reLaunch: h },
        g = function(e) {
          var t = this,
            n = v[e];
          return function(e) {
            if (e) {
              var r = e.currentTarget.dataset,
                o = r.before,
                a = r.after,
                u = r.url;
              try {
                t && o && t[o] && t[o](e);
              } finally {
                u && (n(u), t && a && t[a] && t[a](e));
              }
            }
          };
        },
        y = g("navigate"),
        m = g("redirect"),
        _ = g("switchTab"),
        w = g("reLaunch"),
        b = function(e) {
          void 0 === e && (e = 1), wx.navigateBack({ delta: e });
        },
        $ = function() {
          return getCurrentPages()
            .slice(0)
            .pop();
        },
        O = function(e) {
          var t = /^[\w\-]+(?=\?|$)/u.exec(e);
          return t ? t[0] : u.default.get("nameResolve")(e);
        },
        j = function() {
          var e = $().route;
          return e ? O(e) : "";
        },
        P = function(e) {
          var n = O(e);
          n && o && o.emit("preload:" + n, e, t.queryParse(e.split("?")[1]));
        },
        x = function(e, t) {
          c[e] = t;
        },
        T = function(e) {
          var t = c[e];
          return (c[e] = void 0), t;
        };
      t.default = {
        getPageName: O,
        mount: function(e) {
          var t = e.detail,
            n = f && f(t.id),
            r = n._$ref,
            o = this[t.method];
          switch (t.type) {
            case "attached":
              if (!n) return;
              r && this.$refs && (this.$refs[r] = n), n._$attached(this);
              break;
            case "event:call":
              o && o.apply(this, t.args);
          }
        },
        redirectDelegate: function(e, n) {
          ["navigateTo", "redirectTo", "switchTab", "reLaunch"].forEach(
            function(r) {
              e.on(r, function(e) {
                var o = O(e);
                o && n.emit(r + ":" + o, e, t.queryParse(e.split("?")[1]));
              });
            }
          );
        },
        channel: c,
        dispatcher: function(e) {
          o = e;
        },
        ref: function(e) {
          f = e;
        },
        methods: function(e) {
          (e.$cache = a.default),
            (e.$put = x),
            (e.$take = T),
            (e.$refs = {}),
            (e.$route = s),
            (e.$redirect = d),
            (e.$switch = p),
            (e.$launch = h),
            (e.$back = b),
            (e.$preload = P),
            (e.$curPage = $),
            (e.$curPageName = j),
            (e.$bindRoute = e.$bindNavigate = y),
            (e.$bindRedirect = m),
            (e.$bindSwitch = _),
            (e.$bindReLaunch = w);
        }
      };
    },
    function(e, t, n) {
      "use strict";
      var r =
          (this && this.__spreadArrays) ||
          function() {
            for (var e = 0, t = 0, n = arguments.length; t < n; t++)
              e += arguments[t].length;
            var r = Array(e),
              o = 0;
            for (t = 0; t < n; t++)
              for (var a = arguments[t], u = 0, i = a.length; u < i; u++, o++)
                r[o] = a[u];
            return r;
          },
        o =
          (this && this.__importStar) ||
          function(e) {
            if (e && e.__esModule) return e;
            var t = {};
            if (null != e)
              for (var n in e)
                Object.hasOwnProperty.call(e, n) && (t[n] = e[n]);
            return (t.default = e), t;
          },
        a =
          (this && this.__importDefault) ||
          function(e) {
            return e && e.__esModule ? e : { default: e };
          };
      Object.defineProperty(t, "__esModule", { value: !0 });
      var u = o(n(0)),
        i = a(n(5)),
        c = a(n(1)),
        f = a(n(7)),
        l = a(n(3)),
        s = a(n(4)),
        d = a(n(2)),
        p = new u.default(),
        h = !1,
        v = !1,
        g = !1,
        y = 0,
        m = {
          cache: c.default,
          dispatcher: p,
          redirector: s.default,
          message: u.default,
          channel: i.default.channel
        };
      i.default.ref(f.default.getRef),
        i.default.dispatcher(p),
        f.default.dispatcher(p);
      var _ = function() {
          for (var e = [], t = 0; t < arguments.length; t++)
            e[t] = arguments[t];
          (v = !0), u.message.emit.apply(u.message, r(["app:launch"], e));
        },
        w = function() {
          for (var e = [], t = 0; t < arguments.length; t++)
            e[t] = arguments[t];
          try {
            g ||
              ((g = !0), u.message.emit.apply(u.message, r(["app:show"], e)));
          } finally {
            if (!y) return;
            u.message.emit("app:sleep", new Date().getTime() - y), (y = 0);
          }
        },
        b = function() {
          y = new Date().getTime();
        },
        $ = function(e) {
          var t = e.name || "unknown",
            n = new u.default(),
            o = l.default.get("extendPageBefore");
          if ((o && o(t, e, m), e.onNavigate)) {
            var a = function(t, n) {
              e.onNavigate({ url: t, query: n });
            };
            p.on("navigateTo:" + t, a),
              p.on("redirectTo:" + t, a),
              p.on("switchTab:" + t, a),
              p.on("reLaunch:" + t, a),
              console.log("页面[" + t + "]注册onNavigate");
          }
          e.onPreload &&
            (p.on("preload:" + t, function(t, n) {
              e.onPreload({ url: t, query: n });
            }),
            console.log("页面[" + t + "]注册onPreload")),
            (e.$state = { firstOpen: !1 }),
            (e.$emitter = n),
            i.default.methods(e),
            (e.$on = function(e, t) {
              return p.on(e, t);
            }),
            (e.$emit = function(e) {
              for (var t = [], n = 1; n < arguments.length; n++)
                t[n - 1] = arguments[n];
              return p.emit.apply(p, r([e], t));
            }),
            (e.$off = function(e, t) {
              return p.off(e, t);
            }),
            (e.$ = i.default.mount),
            (e.onLoad = d.default.mergeFun(e.onLoad, function() {
              e.onAwake &&
                u.message.on("app:sleep", function(t) {
                  e.onAwake.call(this, t);
                }),
                h || ((h = !0), e.$state && (e.$state.firstOpen = !0));
            })),
            (e.onReady = d.default.mergeFun(e.onReady, function() {
              s.default.emit("page:ready");
            })),
            e.onPageLaunch && e.onPageLaunch(),
            e.onAppLaunch &&
              (v
                ? e.onAppLaunch.apply(e, v)
                : p.on("app:launch", function(t) {
                    e.onAppLaunch.apply(e, t);
                  })),
            e.onAppShow &&
              (v
                ? e.onAppShow.apply(e, v)
                : p.on("app:show", function(t) {
                    e.onAppShow.apply(e, t);
                  }));
          var c = l.default.get("extendPageAfter");
          return c && c(t, e, m), Page(e), e;
        };
      i.default.redirectDelegate(s.default, p);
      ($.C = f.default),
        ($.A = function(e) {
          if (!e.config || !e.config.route || !e.config.route.length)
            throw new Error("必须配置config.route字段");
          e.config && $.config(e.config);
          var t = e;
          (e.onShow = e.onShow ? d.default.mergeFun(e.onShow, w) : w),
            (e.onHide = e.onHide ? d.default.mergeFun(e.onHide, b) : b),
            (e.onLaunch = e.onLaunch ? d.default.mergeFun(e.onLaunch, _) : _),
            (e.onLaunch = d.default.mergeFun(e.onLaunch, function() {
              t = this;
            })),
            e.onAwake &&
              u.message.on("app:sleep", function(n) {
                e.onAwake.call(t, n);
              }),
            App(e);
        }),
        ($.redirector = s.default),
        ($.message = u.default),
        ($.cache = c.default),
        ($.getPageName = i.default.getPageName),
        ($.config = function(e, t) {
          return (
            "object" == typeof e
              ? Object.keys(e).forEach(function(t) {
                  l.default.set(t, e[t]);
                })
              : l.default.set(e, t),
            this
          );
        }),
        (t.default = $);
    },
    function(e, t, n) {
      "use strict";
      var r =
          (this && this.__assign) ||
          function() {
            return (r =
              Object.assign ||
              function(e) {
                for (var t, n = 1, r = arguments.length; n < r; n++)
                  for (var o in (t = arguments[n]))
                    Object.prototype.hasOwnProperty.call(t, o) && (e[o] = t[o]);
                return e;
              }).apply(this, arguments);
          },
        o =
          (this && this.__spreadArrays) ||
          function() {
            for (var e = 0, t = 0, n = arguments.length; t < n; t++)
              e += arguments[t].length;
            var r = Array(e),
              o = 0;
            for (t = 0; t < n; t++)
              for (var a = arguments[t], u = 0, i = a.length; u < i; u++, o++)
                r[o] = a[u];
            return r;
          },
        a =
          (this && this.__importDefault) ||
          function(e) {
            return e && e.__esModule ? e : { default: e };
          };
      Object.defineProperty(t, "__esModule", { value: !0 });
      var u,
        i = a(n(5)),
        c = a(n(1)),
        f = a(n(3)),
        l = a(n(0)),
        s = a(n(2)),
        d = a(n(4)),
        p = {},
        h = 0,
        v = {
          cache: c.default,
          redirector: d.default,
          message: l.default,
          channel: i.default.channel
        },
        g = function(e) {
          var t = this;
          void 0 === e && (e = {});
          var n = f.default.get("extendComponentBefore");
          n && n(e, v),
            (e.created = s.default.mergeFun(e.created, function() {
              i.default.methods(t);
            })),
            (e.attached = s.default.mergeFun(e.attached, function() {
              var e = (h += 1);
              (t.$id = e),
                (p[e] = t),
                (t._$ref = t.properties.ref || t.properties._ref),
                t.triggerEvent("ing", { id: t.$id, type: "attached" });
            })),
            (e.detached = s.default.mergeFun(e.detached, function() {
              delete p[t.$id];
              var e = t.$parent && t.$parent.$refs,
                n = t._$ref;
              n && e && delete e[n], (t.$parent = null);
            })),
            (e.properties = r(r({}, e.properties), {
              ref: {
                type: String,
                value: "",
                observer: function(e) {
                  if (this._$ref !== e) {
                    var t = this.$parent && this.$parent.$refs;
                    t && (delete t[this._$ref], (this._$ref = e));
                  }
                }
              }
            })),
            (e.methods = r(r({}, e.methods), {
              $set: function() {
                for (var e = [], n = 0; n < arguments.length; n++)
                  e[n] = arguments[n];
                return t.setData.apply(t, e);
              },
              $data: function() {
                return t.data;
              },
              $emit: function(e) {
                for (var t = [], n = 1; n < arguments.length; n++)
                  t[n - 1] = arguments[n];
                return u ? u.emit.apply(u, o([e], t)) : "";
              },
              $on: function(e, t) {
                return u ? u.on(e, t) : "";
              },
              $off: function(e, t) {
                return u ? u.off(e, t) : "";
              },
              $call: function(e) {
                for (var n = [], r = 1; r < arguments.length; r++)
                  n[r - 1] = arguments[r];
                var o = [].slice.call(n, 1);
                t.triggerEvent("ing", {
                  method: e,
                  id: t.$id,
                  type: "event:call",
                  args: o
                });
              },
              _$attached: function(e) {
                (t.$root = e.$root || e), (t.$parent = e);
              },
              $: i.default.mount
            })),
            Component(e);
        };
      (g.getRef = function(e) {
        return p[e];
      }),
        i.default.ref(g.getRef),
        (g.dispatcher = function(e) {
          u = e;
        }),
        (t.default = g);
    }
  ])
);
