// evil-eval/lib/signal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SignalType;
(function (SignalType) {
    SignalType[SignalType["Continue"] = 0] = "Continue";
    SignalType[SignalType["Break"] = 1] = "Break";
    SignalType[SignalType["Return"] = 2] = "Return";
})(SignalType = exports.SignalType || (exports.SignalType = {}));
class Signal {
    constructor(type, value) {
        this.type = type;
        this.value = value;
    }
    static Continue(label) {
        return new Signal(SignalType.Continue, label);
    }
    static Break(label) {
        return new Signal(SignalType.Break, label);
    }
    static Return(value) {
        return new Signal(SignalType.Return, value);
    }
    static isSignal(signal) {
        return signal instanceof Signal;
    }
    static isContinue(signal) {
        return signal instanceof Signal && signal.type === SignalType.Continue;
    }
    static isBreak(signal) {
        return signal instanceof Signal && signal.type === SignalType.Break;
    }
    static isReturn(signal) {
        return signal instanceof Signal && signal.type === SignalType.Return;
    }
}
exports.default = Signal;
