// evil-eval/lib/value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ValueType;
(function (ValueType) {
    ValueType[ValueType["simple"] = 0] = "simple";
    ValueType[ValueType["member"] = 1] = "member";
})(ValueType || (ValueType = {}));
class SimpleValue {
    constructor(value, kind) {
        this.value = value;
        this.kind = kind;
    }
    get v() {
        return this.value;
    }
    set v(value) {
        if (this.kind === 'const') {
            throw new TypeError('Assignment to constant variable');
        }
        else {
            this.value = value;
        }
    }
}
exports.SimpleValue = SimpleValue;
class MemberValue {
    constructor(obj, name) {
        this.obj = obj;
        this.name = name;
    }
    get v() {
        return this.obj[this.name];
    }
    set v(value) {
        this.obj[this.name] = value;
    }
}
exports.MemberValue = MemberValue;
function createSimpleValue(value, kind = 'var') {
    return new SimpleValue(value, kind);
}
exports.createSimpleValue = createSimpleValue;
function createMemberValue(obj, propertyName) {
    return new MemberValue(obj, propertyName);
}
exports.createMemberValue = createMemberValue;
